tests/test_x509_ext.py

```diff
@@ -55,3 +55,104 @@ def test_repr(self):
+        ]
+    )
+    def test_unsupported_types(self, type):
+        with pytest.raises(ValueError):
+            x509.GeneralName(type, "")
+
+    def test_invalid_value_type_rfc822name(self):
+        with pytest.raises(TypeError):
+            x509.GeneralName(x509.GeneralNameType.rfc822_name, 1234)
+
+    def test_invalid_value_type_dnsname(self):
+        with pytest.raises(TypeError):
+            x509.GeneralName(x509.GeneralNameType.dns_name, 1234)
+
+    def test_invalid_value_type_uri(self):
+        with pytest.raises(TypeError):
+            x509.GeneralName(
+                x509.GeneralNameType.uniform_resource_identifier, 1234
+            )
+
+    def test_invalid_value_type_ip_address(self):
+        with pytest.raises(TypeError):
+            x509.GeneralName(x509.GeneralNameType.ip_address, 1234)
+
+    def test_invalid_value_directory_name(self):
+        with pytest.raises(TypeError):
+            x509.GeneralName(x509.GeneralNameType.directory_name, 1234)
+
+    def test_invalid_value_registered_id(self):
+        with pytest.raises(TypeError):
+            x509.GeneralName(x509.GeneralNameType.registered_id, 1234)
+
+    def test_repr(self):
+        gn = x509.GeneralName(x509.GeneralNameType.dns_name, "cryptography.io")
+        assert repr(gn) == (
+            "<GeneralName(type=GeneralNameType.dns_name, value=cryptography.io"
+            ")>"
+        )
+
+    def test_eq(self):
+        assert x509.GeneralName(
+            x509.GeneralNameType.dns_name, "cryptography.io"
+        ) == x509.GeneralName(x509.GeneralNameType.dns_name, "cryptography.io")
+
+    def test_ne(self):
+        assert x509.GeneralName(
+            x509.GeneralNameType.dns_name, "cryptography.io"
+        ) != x509.GeneralName(
+            x509.GeneralNameType.uniform_resource_identifier, "cryptography.io"
+        )
+
+        assert x509.GeneralName(
+            x509.GeneralNameType.dns_name, "cryptography.io"
+        ) != x509.GeneralName(x509.GeneralNameType.dns_name, "ryptography.io")
+
+        assert x509.GeneralName(
+            x509.GeneralNameType.dns_name, "cryptography.io"
+        ) != object()
+
+
+class TestSubjectAlternativeName(object):
+    # TODO bad name
+    def test_dns(self):
+        san = x509.SubjectAlternativeName(
+            [
+                x509.GeneralName(
+                    x509.GeneralNameType.dns_name, "cryptography.io"
+                )
+            ]
+        )
+        names = san.get_values_for_type(x509.GeneralNameType.dns_name)
+        assert names == ["cryptography.io"]
+
+    def test_iter_names(self):
+        gns = [
+            x509.GeneralName(x509.GeneralNameType.dns_name, "cryptography.io"),
+            x509.GeneralName(
+                x509.GeneralNameType.dns_name, "www.cryptography.io"
+            ),
+        ]
+        san = x509.SubjectAlternativeName(gns)
+        assert len(san) == len(gns)
+        # This assert sucks. It just does an identity test so __eq__ never gets
```

Comment:
My brain is fried so this stayed in this initial PR. Suggestions on how to do a better test here welcome.